docs/markdown/Builtin-options.md

```diff
@@ -347,3 +348,13 @@ installation is a virtualenv, and use `venv` or `system` as appropriate (but
 never `prefix`). This option is mutually exclusive with the `platlibdir`/`purelibdir`.
 
 For backwards compatibility purposes, the default `install_env` is `prefix`.
+
+*Since 1.1.0* The `python.bytecompile` option can be used to enable compiling
+python bytecode. Bytecode has 3 optimization levels:
+
+- 0, bytecode without optimizations
+- 1, bytecode with some optimizations
+- 2, bytecode with some more optimizations
+
+To this, Meson adds level `-1`, which is to not attempt to compile bytecode at
+all.
```

Comment:
I'm not a big fan of `-1`, `none` feels more natural. I guess that would mean having a string option, but from CLI that makes no difference. Also, shouldn't the default be `-1`?

Reply:
> Instead of the possible values being: [none, 0, 1, 2]

Matter of taste I guess, but I prefer that notation in this case.

> I vacillated between that and 0 for a while

What make me doubt here is how it will play with distro. Debian doesn't ship pyc, suddenly having pyc in their destdir could be an issue? Likewise other distro likely have their own compilation tools which could conflict with meson doing it as well? Making it opt-in is the safest solution. But I agree it is a dilemma.

I guess distro have a global list of Meson options they can disable in a single place for all packages?

Reply:
Many distros have that, yeah. Including Debian. :)

This is actually a pretty good topic to discuss, so let me lay out the options:
- distro wants bytecode and ignores what we do, generates their own
- distro is happy to let us decide
- distro doesn't want the bytecode we made, at all


In case 1, we're fine. The bytecode we make doesn't conflict with the "even more" that they add afterwards.

In case 2, we're also fine, by definition, as long as we have a good default. We're saving them the trouble, even.

Case 3 is the interesting one, so let's discuss that. I'm aware of two distros that are like this:
- voidlinux, currently allows python packages to install their own bytecode and makes no effort to stop them. Instead every python packages automatically deletes all bytecode, then adds a rule to generate it after installation, on the user's machine, at level =1
- Debian, has lintian rules that warn you if you have bytecode. For popular build systems, debhelper automatically passes the relevant flags to disable it (but debhelper's meson plugin doesn't handle this option that I'm adding in this PR). A rule is added to generate it after installation, on the user's machine, at level =0

Reply:
What exactly does the default here do? Which level does it pick? Iirc pypa/installer by default compiles regular, -O and -OO bytecode.

Reply:
I disagree. We have hit end-users ending up using optimized bytecode. It's problematic when you're generating bytecode into read-only storage. Then said user will end up regenerating bytecode every time their program starts but may not notice it without checking for debug output.

Reply:
Another axis to consider here is timestamped vs checked hash vs unchecked hash files. By default when SOURCE_DATE_EPOCH is set, compileall creates checked hash. This may or may not be what distro intends.

Reply:
Under what circumstances do you hit this? Like I said, having the option is important -- I'm just unsure it's a good default to optimize for.

The downside of not generating the optimized bytecode users end up using, is slower startup because it cannot be cached to disk.

The downside of generating optimized bytecode users never end up using, is potentially tripling storage requirements.

What context is this read-only storage? Yocto? Custom repositories for a linux package manager? Running pip install and freezing the results into a squashfs? Something else?

Reply:
Close. Flatpak runtimes which get shipped over ostree and mounted as immutable. That said, any ostree-based distro would probably do the same.

Reply:
It's not a big deal. Worst-case Meson-generated bytecode can be simply replaced but I would hope this to be configurable enough that wouldn't be needed.

Reply:
Ok, I think we can close this review comment.